spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -212,6 +216,17 @@ abstract class Graph[T: ClassTag](
         || inputNames.contains(n.element.getName()))
     backwardTargets.foreach(_ -> dummyBackwardEnd)
     backwardGraph = dummyBackwardEnd.graph(true)
+
+    // Check if gradInput is empty for each input
+    isGradInputAvailable = inputs.map(_ => false).toArray
+    inputs.zipWithIndex.map{ case (n, i) =>
+      backwardGraph.DFS.foreach(curNode => {
+        if (curNode.element.getName() == n.element.getName() && !isStopGradient(n.element)) {
+          isGradInputAvailable(i) = true
+        }
+      })
```

Comment:
Maybe it does not matter that much, but I think we are able to do this in one DFS pass.